src/utils/hierarchical_model.py

```diff
@@ -0,0 +1,213 @@
+
+    Args:
+        data (pd.DataFrame): The dataset to draw the dendrogram plot of.
+        method_list (list): The linkage methods to use. Defaults to ['single', 'complete', 'average', 'weighted', 'centroid', 'median', 'ward'].
+        distance_metric (str): The distance metric to use when calculating the linkage matrix. Defaults to 'euclidean'.
+        optimal_ordering (bool): Whether to use optimal leaf ordering when drawing the dendrogram plot. Defaults to False.
+    """
+
+    # Set default linkage methods if not provided
+    if method_list is None:
+        method_list = ['single', 'complete', 'average', 'weighted', 'centroid', 'median', 'ward']
+
+    # Create a figure to plot the dendrograms
+    fig, axs = plt.subplots(len(method_list), 1, figsize=(10, 30), sharex=True, gridspec_kw={'hspace': 0.5})
+
+    # Loop over the methods and plot the corresponding dendrogram
+    for index, method in enumerate(method_list):
+        # Calculate the linkage matrix
+        linkage_matrix = get_linkage(data, method, distance_metric, optimal_ordering)
+        
+        # Plot the dendrogram
+        dendrogram(linkage_matrix, ax=axs[index])
+        
+        # Set the plot title and axis labels
+        axs[index].set_title(f'Dendrogram ({method} linkage)')
+        axs[index].set_xlabel('Sample Index')
+        axs[index].set_ylabel('Distance')
+    
+    # Show the plot and return the dendrogram object
+    plt.show()
+
+
+def plot_thresholded_dendrogram(linkage_matrix: ndarray, distance_threshold: int) -> None:
+    """Plots the dendrogram with a horizontal line at the given threshold.
+
+    Args:
+        linkage_matrix (numpy.ndarray): The linkage matrix of the dataset.
+        threshold (float): The threshold distance to plot the horizontal line at.
+    """
+    # Plot the dendrogram
+    plt.figure(figsize=(10, 7))
+    dendrogram(linkage_matrix)
+    
+    # Set the plot title and axis labels
+    plt.title("Dendrogram with Threshold")
+    
+    # Draw the horizontal line
+    plt.axhline(y=distance_threshold, color='black', linestyle='--')
+    
+    # Show the plot
+    plt.show()
+
+
+def hierarchical_clustering(data: pd.DataFrame, k=3) -> ndarray:
+    """To cluster one-hot encoded data using fcluster from hierarchical clustering algorithms.
+
+    Args:
+        linkage_matrix (ndarray): The linkage matrix of the dataset.
+        k (int): number of clusters.
+        criterion (str, optional): The criterion to use in forming flat clusters. Defaults to 'maxclust'.
+
+    Returns:
+        ndarray: An array of length n. T[i] is the flat cluster number to which original observation i belongs.
+    """
+    return fclusterdata(data, t=k, criterion='maxclust', method='ward')
+
+
+def _map_color_cluster(labels: ndarray) -> pd.Series(str):
+    """Assigns a color to each cluster in the given DataFrame based on their label.
+    Args:
+        df (pd.DataFrame): The DataFrame containing the cluster labels.
```

Comment:
There is no dataframe, u did not attach "labels" to the documentation.

Reply:
I use this function in several places in the notebooks.

Reply:
I fixed it

Reply:
Fixed it.